File: LeetCode/0002_add-two-numbers/20220519_iteration.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l = ListNode()
        p = l
        carry = 0
        
        while l1 or l2 or p:
            if l1 == None:
                v1 = 0
            else:
                v1 = l1.val
                l1 = l1.next
                
            if l2 == None:
                v2 = 0
            else:
                v2 = l2.val
                l2 = l2.next
                
            s = p.val + v1 + v2

            carry = s // 10
            s %= 10
            
            p.val = s
            if carry > 0:
                p.next = ListNode(val=carry)
            elif l1 or l2:
                p.next = ListNode()
                
            p = p.next
            
        return l

File: LeetCode/0002_add-two-numbers/20220519_iteration.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val l = ListNode(0)
        var lp: ListNode? = l
        
        var lp1 = l1
        var lp2 = l2
        
        while (lp != null) {
            val v1 = lp1?.`val` ?: 0
            val v2 = lp2?.`val` ?: 0
            
            lp1 = lp1?.next
            lp2 = lp2?.next
            
            val sum = lp.`val` + v1 + v2
            
            lp.`val` = sum % 10
            val carry = sum / 10
            
            if (lp1 != null || lp2 != null || carry > 0) {
                lp.next = ListNode(carry)
            }
            
            lp = lp.next
        }
        
        return l
    }
}